### PR TITLE
[Distance metrics M1] support for L2 distance

### PR DESCRIPTION
**Pull request by tjgreen42**
## Description
This PR adds support for distance metrics besides `cosine`, beginning with `L2`.  The primitives for this metric were already present (earlier versions of `pgvectorscale` used `L2` rather than cosine), so this PR just adds the operator class-related plumbing.  As discussed in the design doc (see references) we follow `pgvector` syntax.

## Testing
* Refactored some of the basic scaffolds to take a `DistanceType` parameter and added `L2` versions of some of the callers.  We could carry this out systemically and insist that all tests have variants covering all metrics, but this seems like overkill.
* Confirmed manually via `EXPLAIN PLAN` that the correct index is still chosen.
* Confirmed manually that `test_upgrade` succeeds (with generated sql schema, not included in this PR)

## References
* https://timescale.slab.com/posts/distance-metrics-for-pgvectorscale-atrovg6n?shr=NaR3yhZgvzEx6SixcBHwKsqV
* https://github.com/timescale/pgvectorscale/issues/95
* https://linear.app/timescale/issue/AI-194/pgvectorscale-add-l2-inner-product-l1-metrics

### CODE REVIEW

File: pgvectorscale/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "vectorscale"
-version = "0.4.0"
+version = "0.5.0"
```
Comment (syvb) [inline pgvectorscale/Cargo.toml:3]: I think it's better to have a separate PR (e.g. #140 for 0.4.0) for updating the version.

Comment (cevian) [inline pgvectorscale/Cargo.toml:3]: we usually make this change during release and not in PRs

File: pgvectorscale/sql/vectorscale--0.4.0--0.5.0.sql [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/* <begin connected objects> */
```
Comment (cevian) [inline pgvectorscale/sql/vectorscale--0.4.0--0.5.0.sql:1]: we usually add this file during release and not in PRs. Why was this needed now? for testing?

File: pgvectorscale/src/access_method/mod.rs [hunk 3]
```diff
@@ -99,16 +102,30 @@ BEGIN
     AND c.opcmethod = (SELECT oid FROM pg_catalog.pg_am am  WHERE am.amname = 'diskann');
 
     IF c = 0 THEN
-        CREATE OPERATOR CLASS vector_cosine_ops DEFAULT
+        CREATE OPERATOR CLASS vector_cosine_ops
```
Comment (cevian) [inline pgvectorscale/src/access_method/mod.rs:105]: I think this needs to be a bit more involved. There is now a chance that vector_cosine_ops exists but without the Function 1. In that case, we need to drop the old ops and recreate. We should try to keep this logic idempotent.

Comment (cevian) [reply pgvectorscale/src/access_method/mod.rs:105]: I would also investigate why the tests are passing. The upgrade test should fail now. Is it not being run?

**Review by cevian — CHANGES_REQUESTED** [review]
Overall looks good but some comments

File: pgvectorscale/src/access_method/mod.rs (hunk 3 above)
Comment (syvb) [reply pgvectorscale/src/access_method/mod.rs:105]: @cevian The upgrade test [is `#[ignore]`d](https://github.com/timescale/pgvectorscale/blob/9802ac91db55cfea05eb10b717b13fe736eb101c/pgvectorscale/src/access_method/upgrade_test.rs#L25-L27), and only run manually

File: pgvectorscale/src/access_method/mod.rs [hunk 4]
```diff
@@ -88,27 +91,23 @@ fn amhandler(_fcinfo: pg_sys::FunctionCallInfo) -> PgBox<pg_sys::IndexAmRoutine>
...
-    AND c.opcmethod = (SELECT oid FROM pg_catalog.pg_am am  WHERE am.amname = 'diskann');
-
-    IF c = 0 THEN
-        CREATE OPERATOR CLASS vector_cosine_ops DEFAULT
-        FOR TYPE vector USING diskann AS
-	        OPERATOR 1 <=> (vector, vector) FOR ORDER BY float_ops;
-    END IF;
-END;
-$$;
+    DROP OPERATOR CLASS IF EXISTS vector_cosine_ops USING diskann;
```
Comment (cevian) [inline pgvectorscale/src/access_method/mod.rs:94]: I am doubtful its safe/good to always drop the class like this. I'd rather do it if its missing or missing the function. Also what happens if an index already exists and we try dropping the opclass? Wouldn't the drop fail if no CASCADE? And the CASCADE would drop the index. This seem problematic.

**Review by cevian — CHANGES_REQUESTED** [review]
Everything looks good except for the update script which I still have concerns about.

**Review by syvb — APPROVED** [review]

**Review by cevian — APPROVED** [review]
may make sense to squash some commits
